pkg/controllers/jobset_controller.go

```diff
@@ -603,31 +607,209 @@ func executeSuccessPolicy(ctx context.Context, js *jobset.JobSet, ownedJobs *chi
+			if i < numRules {
+				rule = ptr.To(rules[i])
+			}
+			if ruleIsApplicable(rule, failedJob, jobFailureReason) && (buckets[i].firstFailureTime == nil || jobFailureTime.Before(buckets[i].firstFailureTime)) {
+				buckets[i].firstFailedJob = failedJob
+				buckets[i].firstFailureTime = jobFailureTime
+			}
+		}
+	}
+
+	// Checking if any failure policy rules were matched
+	// and returning the rule along with the first
+	// failed job to match it.
+	for i := 0; i < numRules; i++ {
+		if buckets[i].firstFailedJob != nil {
+			return &rules[i], buckets[i].firstFailedJob
+		}
+	}
+
+	// We get here when no rule matched any of the failure policy rules
+	return nil, buckets[defaultRuleIndex].firstFailedJob
+}
+
+// parentReplicatedJobName returns the name of the parent
+// ReplicatedJob and true if it is able to retrieve the parent.
+// The empty string and false are returned otherwise.
+func parentReplicatedJobName(job *batchv1.Job) (string, bool) {
+	if job == nil || job.Labels == nil {
+		return "", false
+	}
+
+	replicatedJobNameKey := job.Labels[jobset.ReplicatedJobNameKey]
+	replicatedJobNameIsUnset := replicatedJobNameKey == ""
+	return replicatedJobNameKey, !replicatedJobNameIsUnset
+}
+
+// If the rule is not nil, then the functions returns:
+// - true if the rule is applicable to the jobFailureReason;
+// - false other.
+// If the rule is nil, we return true.
+func ruleIsApplicable(rule *jobset.FailurePolicyRule, failedJob *batchv1.Job, jobFailureReason string) bool {
+	if rule == nil {
+		return true
+	}
+
+	ruleAppliesToJobFailureReason := len(rule.OnJobFailureReasons) == 0 || slices.Contains(rule.OnJobFailureReasons, jobFailureReason)
+	if !ruleAppliesToJobFailureReason {
+		return false
+	}
+
+	parentReplicatedJob, exists := parentReplicatedJobName(failedJob)
+	if !exists {
+		// If we cannot find the parent ReplicatedJob, we assume the rule does not apply.
+		return false
+	}
+
+	ruleAppliesToParentReplicatedJob := len(rule.TargetReplicatedJobs) == 0 || slices.Contains(rule.TargetReplicatedJobs, parentReplicatedJob)
+	return ruleAppliesToParentReplicatedJob
+}
+
+// The type failurePolicyActionApplier applies a FailurePolicyAction and returns nil if the FailurePolicyAction was successfully applied.
+// The function returns an error otherwise.
+type failurePolicyActionApplier = func(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOpts *statusUpdateOpts) error
+
+// -------------------------Helper Functions for Each FailurePolicyActions Start------------------------
+
+// failJobSetActionApplier applies the FailJobSet FailurePolicyAction
+var failJobSetActionApplier failurePolicyActionApplier = func(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOpts *statusUpdateOpts) error {
+	setJobSetFailedCondition(ctx, js, constants.ReachedMaxRestartsReason, messageWithFirstFailedJob(constants.ReachedMaxRestartsMessage, matchingFailedJob.Name), updateStatusOpts)
```

Comment:
nit: can we store `messageWithFirstFailedJob(constants.ReachedMaxRestartsMessage, matchingFailedJob.Name)` in a variable and use it to make this line `setJobSetFailedCondition(...)` not so long?

Right now at first it kind of looks like setJobSetFailedCondition() is called and then messageWithFirstFailedJob() is called after, until I look more closely and notice the 2nd call is a large inline call used as a parameter in the 1st. 

Something like this would be clearer at a glance I think: 
```go
msg := messageWithFirstFailedJob(constants.ReachedMaxRestartsMessage, matchingFailedJob.Name)
setJobSetFailedCondition(ctx, js, constants.ReachedMaxRestartsReason, msg, updateStatusOpts)
```

Reply:
This is done. A good nit IMHO.